### PR TITLE
JP2K Improvments

### PR DESCRIPTION
**Pull request by rotemtzaban**
Iv'e made performance improvements to the direct JP2K direct when using kakadu, mainly creating only one imageReader for every JP2KReader when using kakadu.
Since the kakadu imageReader is thread safe, it is possible to use one reader, and creating a reader per request was unnecessary, and caused a large performance decrease.
The other readers however aren't thread safe, so i needed to create a reader per request depending on the reader available.

### CODE REVIEW

**Comment by bencaradocdavies** [comment]
Please also squash into a single commit. I find merge commits in pull requests hard to follow.

File: modules/plugin/jp2k/src/main/java/org/geotools/coverageio/jp2k/Granule.java [hunk 1]
```diff
@@ -318,13 +269,15 @@ public RenderedImage loadRaster(
...
-				raster= request.getReadType().read(readParameters,imageIndex, granuleFile, selectedlevel.rasterDimensions,tileDimension,cachedSPI);
+				raster= request.getReadType().read(readParameters,imageIndex, granuleFile, selectedlevel.rasterDimensions,tileDimension,cachedSPI, imageReader);
 				if (raster == null)
 					return null;
 			}
 			catch (Throwable e) {
-				if(LOGGER.isLoggable(java.util.logging.Level.FINE))
-					LOGGER.log(java.util.logging.Level.FINE,"Unable to load raster for granule "+this.toString()+ " with request "+request.toString(),e);
+                if (LOGGER.isLoggable(java.util.logging.Level.FINE)) {
+                    LOGGER.log(java.util.logging.Level.FINE,"Unable to load raster for granule "+this.toString()+
```
Comment (bencaradocdavies) [inline modules/plugin/jp2k/src/main/java/org/geotools/coverageio/jp2k/Granule.java]: Whitespace around operators, please.

File: modules/plugin/jp2k/src/main/java/org/geotools/coverageio/jp2k/Granule.java [hunk 2]
```diff
@@ -470,29 +390,14 @@ public Level getLevel(final int index) {
 					final Level newLevel=new Level(scaleX,scaleY,levelDimension.width,levelDimension.height);
 					this.granuleLevels.put(Integer.valueOf(index),newLevel);
 					return newLevel;
-					
-
-				} catch (IllegalStateException e) {
-					throw new IllegalArgumentException(e);
-					
-				} catch (IOException e) {
+				} catch (IllegalStateException|IOException e) {
```
Comment (bencaradocdavies) [inline modules/plugin/jp2k/src/main/java/org/geotools/coverageio/jp2k/Granule.java]: Whitespace around operators (even though the style guide predates this syntax).

**Comment by tbarsballe** [comment]
Thanks for the contribution.
Since this appears to be your first contribution to GeoTools, could you please sign a Contributor License Agreement?
Also, is there a GEOT issue for this in the [OSGeo JIRA](https://osgeo-org.atlassian.net/browse)? If not, could you please create one and update the title of this PR, and commit messages with the ticket id.

See [CONTRIBUTING.md](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) for more details.

File: modules/plugin/jp2k/src/main/java/org/geotools/coverageio/jp2k/RasterLayerResponse.java [hunk 3]
```diff
@@ -138,154 +124,71 @@ public int getImageIndex() {
...
-						LOGGER.fine("Loading image number " + granuleIndex+ " failed, original request was "+request);
-					continue;
+			Color inputTransparentColor = request.getInputTransparentColor();
+			boolean doInputTransparency = inputTransparentColor != null;
+
+			final RenderedImage loadedImage;
+			try {
+				loadedImage=loader.loadGranule();
+				if(loadedImage==null)
+				{
```
Comment (bencaradocdavies) [inline modules/plugin/jp2k/src/main/java/org/geotools/coverageio/jp2k/RasterLayerResponse.java]: ```{``` on the same line as ```if```.

File: modules/plugin/jp2k/src/main/java/org/geotools/coverageio/jp2k/RasterLayerResponse.java [hunk 4]
```diff
@@ -138,154 +124,71 @@ public int getImageIndex() {
...
+				LOGGER.log(Level.SEVERE,"Unable to load the raster" ,e);
+				return;
 			}
-
-			granulesNumber=granuleIndex;
-			if(granulesNumber==0)
-			{
-				if(LOGGER.isLoggable(Level.FINE))
-					LOGGER.log(Level.FINE,"Unable to load any data ");
+			catch (ImagingException|javax.media.jai.util.ImagingException  e) {
```
Comment (bencaradocdavies) [inline modules/plugin/jp2k/src/main/java/org/geotools/coverageio/jp2k/RasterLayerResponse.java]: ```catch``` on the same line as ```}```

File: modules/plugin/jp2k/src/main/java/org/geotools/coverageio/jp2k/RasterManager.java [hunk 5]
```diff
@@ -443,15 +444,14 @@ public RasterManager(final JP2KReader reader) throws DataSourceException {
         overviewsController=new OverviewsController();
         decimationController= new DecimationController();
         try {
-			spatialDomainManager= new SpatialDomainManager();
-		} catch (TransformException e) {
-			throw new DataSourceException(e);
-		} catch (FactoryException e) {
+        	spatialDomainManager= new SpatialDomainManager();
```
Comment (bencaradocdavies) [inline modules/plugin/jp2k/src/main/java/org/geotools/coverageio/jp2k/RasterManager.java]: No tabs in source code, please.

File: modules/plugin/jp2k/src/main/java/org/geotools/coverageio/jp2k/RasterManager.java [hunk 6]
```diff
@@ -498,23 +497,22 @@ private OverviewPolicy extractOverviewPolicy() {
...
-		
-	}
+	public void dispose() {	}
 
 	public URL getInputURL() {
 		return inputURL;
 	}
 
+	public Granule getGranule() {
+		return granule;
```
Comment (bencaradocdavies) [inline modules/plugin/jp2k/src/main/java/org/geotools/coverageio/jp2k/RasterManager.java]: No tabs, please.

File: modules/plugin/jp2k/src/main/java/org/geotools/coverageio/jp2k/ReadType.java [hunk 7]
```diff
@@ -92,90 +82,25 @@ RenderedImage read(
...
-    				if (LOGGER.isLoggable(Level.WARNING))
-    					LOGGER.warning("Unable to get reader for file "
-    							+ rasterFile.getAbsolutePath());
-    				return null;
-    			}
-    			
-    			inStream.reset();
-    			reader.setInput(inStream);
+			if(CoverageUtilities.checkEmptySourceRegion(readP, readDimension))
+				return null;
```
Comment (bencaradocdavies) [inline modules/plugin/jp2k/src/main/java/org/geotools/coverageio/jp2k/ReadType.java]: No tabs, please.

Comment (bencaradocdavies) [inline modules/plugin/jp2k/src/main/java/org/geotools/coverageio/jp2k/ReadType.java]: And curly braces around all if clauses. Local reformats may be necessary to fix tabs and for local consistency.

**Comment by aaime** [comment]
@dromagnoli is going to have a look at this

**Comment by dromagnoli** [comment]
Thanks for the contribution.
I agree with previous comments made by Ben and Torben:
Do you have a GEOT Jira for this? 
Can you squash the commits?
When you add some code, please make sure to avoid tabs and also make sure to have catch on the same line of the previous closing curly brace and similar rules apply on the IF blocks. 
You may want to take a look at the Eclipse formatter template:
https://github.com/geotools/geotools/blob/master/build/eclipse/formatter.xml
Just make sure to avoid a blind reformat of the whole code in a shot in order to make the PR still readable. 

I have also some additional comments:
In general, although some old code may have ugly formatting or indentations at the beginning, please just reformat the pieces of code that you are modifying otherwise the Pull Request becomes huge and it is "hard" to read/review when looking for real changes at the code. (No need to revert all these changes at this point).

I see that you have removed the Guard logs checks.
That should be ok when the code was simply logging a trivial message, as an instance logger.fine("Unable to load any data") or similar. However, I guess that the guarded logging should be present when the logging involves some string concatenation or object composition.

**Comment by aaime** [comment]
The idea is good, but there was no response to the feedback and no CLA being signed. If someone wants it, we'll have to rewrite it from scratch. So, closing it.
